src/common/core/mixincore.js

```diff
@@ -0,0 +1,817 @@
+], function (ASSERT, TASYNC, CANON, CONSTANTS) {
+    'use strict';
+
+    var MixinCore = function (innerCore, options) {
+        ASSERT(typeof options === 'object');
+        ASSERT(typeof options.globConf === 'object');
+        ASSERT(typeof options.logger !== 'undefined');
+
+        var logger = options.logger,
+            key,
+            self = this;
+
+        for (key in innerCore) {
+            this[key] = innerCore[key];
+        }
+
+        logger.debug('initialized MixinCore');
+
+        //<editor-fold=Helper Functions>
+
+        function realNode(node) { //TODO we have to make some more sophisticated distinction
+            if (self.getPath(node).indexOf('_') !== -1) {
+                return false;
+            }
+            return true;
+        }
+
+        function getOrderedMixinList(node) {
+            var paths = self.getMixinPaths(node),
+                metaNodes = self.getAllMetaNodes(node),
+                helper = {},
+                orderedList = [],
+                guid,
+                i;
+
+            for (i = 0; i < paths.length; i += 1) {
+                if (metaNodes[paths[i]]) {
+                    guid = self.getGuid(metaNodes[paths[i]]);
+                    helper[guid] = paths[i];
+                    orderedList.push(guid);
+                }
+
+            }
+
+            orderedList.sort();
+
+            for (i = 0; i < orderedList.length; i += 1) {
+                orderedList[i] = metaNodes[helper[orderedList[i]]];
+            }
+
+            return orderedList;
+        }
+
+        function getNodeDictionary(node, paths) {
+            var allMetaNodes = self.getAllMetaNodes(node),
+                dictionary = {},
+                i;
+
+            for (i = 0; i < paths.length; i += 1) {
+                if (allMetaNodes[paths[i]]) {
+                    dictionary[paths[i]] = allMetaNodes[paths[i]];
+                }
+            }
+
+            return dictionary;
+        }
+
+        function extendUniqueArray(base, extension) {
+            var i;
+
+            for (i = 0; i < extension.length; i += 1) {
+                if (base.indexOf(extension[i]) === -1) {
+                    base.push(extension[i]);
+                }
+            }
+        }
+
+        function getExtendedUniqueKeyedObject(base, extension) {
+            var extended = {},
+                key;
+            for (key in base) {
+                extended[key] = base[key];
+            }
+
+            for (key in extension) {
+                if (!base[key]) {
+                    extended[key] = extension[key];
+                }
+            }
+
+            return extended;
+        }
+
+        function getValidNames(node, getter, alreadyVisited) {
+            var base = self.getBase(node),
+                guid = self.getGuid(node),
+                names,
+                mixins = getOrderedMixinList(node),
+                i;
+
+            if (alreadyVisited.indexOf(guid) !== -1) {
```

Comment:
good point, I'll make a note about it, but currently we are not expecting huge graphs of nodes...